ocaml/xapi/helpers.ml

```diff
@@ -1184,6 +1184,15 @@ let retry ~__context ~doc ?(policy = Policy.standard) f =
 let retry_with_global_lock ~__context ~doc ?policy f =
   retry ~__context ~doc ?policy (fun () -> with_global_lock f)
 
+(* Retry function f until success or timeout, f should return true on success *)
+let rec retry_until_timeout ?(interval=0.1) ?(timeout=2.) doc f =
+  if timeout <= 0.0 then raise Api_errors.(Server_error(internal_error,
+                                                        [Printf.sprintf "retry %s failed with interval %f and timeout %f" doc interval timeout]));
+  match f() with
+  | true -> ()
+  | false -> Thread.delay interval;
+    retry_until_timeout ~interval ~timeout:(timeout -. interval) doc f
```

Comment:
You could use `interval *. 1.5` to increasingly slow down the loop but use a longer timeout of 5s.

Reply:
sounds interesting, it will have more check density at beging while fewer check density later.
this possible the expected behavior

Reply:
updated, please have a look